cmd/shim/main.go

```diff
@@ -322,3 +347,87 @@ func toggleNesting(ctx context.Context) ([]string, error) {
 	}
 	return []string{}, nil
 }
+
+type SecretScrubWriter struct {
+	mu            sync.Mutex
+	w             io.Writer
+	secretToScrub core.SecretToScrubInfo
+	secretValues  []string
+	fs            fs.FS
+}
+
+func (w *SecretScrubWriter) Write(b []byte) (int, error) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	s := string(b)
+	for _, secret := range w.secretValues {
+		// FIXME: I think we can do better
+		s = strings.ReplaceAll(s, secret, "***")
+	}
+
+	n, err := w.w.Write([]byte(s))
+	if err != nil {
+		return -1, err
+	}
+	if n != len(b) {
+		n = len(b)
+	}
+
+	return n, err
+}
+
+// NewSecretScrubWriter replaces known secrets by "***".
+// The value of the secrets referenced in secretsToScrub are loaded either
+// from env or from the fs.
+func NewSecretScrubWriter(w io.Writer, fsys fs.FS, env []string, secretsToScrub core.SecretToScrubInfo) (io.Writer, error) {
+	secrets := loadSecretsToScrubFromEnv(env, secretsToScrub.Envs)
+
+	fileSecrets, err := loadSecretsToScrubFromFiles(fsys, secretsToScrub.Files)
+	if err != nil {
+		return nil, fmt.Errorf("could not load secrets from file: %w", err)
+	}
+	secrets = append(secrets, fileSecrets...)
+
+	return &SecretScrubWriter{
+		fs:            fsys,
+		w:             w,
+		secretValues:  secrets,
+		secretToScrub: secretsToScrub,
+	}, nil
+}
+
+func loadSecretsToScrubFromEnv(env []string, secretsToScrub []string) []string {
+	secrets := []string{}
+
+	for _, envKV := range env {
+		envName, envValue, ok := strings.Cut(envKV, "=")
+		// no env value for this secret
+		if !ok {
+			continue
+		}
+
+		for _, envToScrub := range secretsToScrub {
+			if envName == envToScrub {
+				secrets = append(secrets, envValue)
+			}
+		}
+	}
+
+	return secrets
+}
+
+func loadSecretsToScrubFromFiles(fsys fs.FS, secretFilePathsToScrub []string) ([]string, error) {
+	secrets := make([]string, 0, len(secretFilePathsToScrub))
+
+	for _, fileToScrub := range secretFilePathsToScrub {
+		// we remove the first `/` from the fileToScrub to work with fs.ReadFile
+		secret, err := fs.ReadFile(fsys, fileToScrub[1:])
```

Comment:
Secret file mounts can also be relative to the current working directory, so I don't think you want to lop off the first character:

```go

	secret := c.Directory().
		WithNewFile("/secret", "some-content").
		File("/secret").
		Secret()

	out, err := c.Container().
		From("alpine:3.16.2").
		WithWorkdir("/anything").
		WithMountedSecret("sekret", secret).
		WithExec([]string{"cat", "sekret"}).
		Stdout(ctx)
	require.NoError(t, err)
	require.Equal(t, "some-content", out)
```

You could call `filepath.Abs` on `fileToScrub` to get an absolute path first.

Reply:
Good point. Thanks for that info! Will fix that.

Reply:
done

Reply:
I don't see the fix for this - maybe forgot to push? :)

Reply:
Nope, somehow I thought I fixed it.

Reply:
done: https://github.com/dagger/dagger/pull/4518/commits/adca5e43871d44fdefaa95b4b0171a005e25169c